core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -465,29 +465,64 @@ trait EntityComponent {
+            // since the size of the list has decreased, delete the extra rows from table
+            val deleteIds = existingAttrsToRecordIds(name).toSeq.takeRight(existingAttrSize - updateAttrSize)
+            (Seq.empty[EntityAttributeRecord], attrRecords, deleteIds)
+          }
+          else (Seq.empty[EntityAttributeRecord], attrRecords, Seq.empty[Long]) // the list size hasn't changed
+        }
+
+        def recordsForUpdateAttribute(name: AttributeName, attribute: Attribute, attrRecords: Seq[EntityAttributeRecord]) = {
+          attribute match {
+            case list: AttributeList[_] => checkAndUpdateRecsForListAttr(name, list, attrRecords)
+            case _ => (Seq.empty[EntityAttributeRecord], attrRecords, Seq.empty[Long])
+          }
+        }
+
         lookupNotYetLoadedReferences(workspaceContext, entityRefsToLookup, Seq(entityRecord.toReference)) flatMap { entityRefRecs =>
           val allTheEntityRefs = entityRefRecs ++ Seq(entityRecord) //re-add the current entity
           val refsToIds = allTheEntityRefs.map(e => e.toReference -> e.id).toMap
           val existingAttributes = existingAttrsToRecordIds.keys.toSeq
 
-          val (updateAttrs, insertAttrs) = upserts.partition( attr => existingAttributes.contains(attr._1) )
-
-          //function that marshals attribute maps to list of attribute records for saving
-          def attributeMapToRecs(attrMap: AttributeMap): Iterable[EntityAttributeRecord] = {
-            for {
-              (attributeName, attribute) <- attrMap
-              attributeRec <- entityAttributeQuery.marshalAttribute(refsToIds(entityRecord.toReference), attributeName, attribute, refsToIds)
-            } yield attributeRec
-          }
-
-          //actually build the records we're going to send to the db
-          val insertRecs = attributeMapToRecs(insertAttrs)
-          val updateRecs = attributeMapToRecs(updateAttrs)
+          //get ids that need to be deleted from db
           val deleteIds = (for {
             attributeName <- deletes
-          } yield existingAttrsToRecordIds.get(attributeName)).flatten
+          } yield existingAttrsToRecordIds.get(attributeName)).flatten.flatten
+
+          val (insertRecs: Seq[EntityAttributeRecord], updateRecs: Seq[EntityAttributeRecord], extraDeleteIds: Seq[Long]) = upserts.map {
+            case (name, attribute) =>
+              val attrRecords = entityAttributeQuery.marshalAttribute(refsToIds(entityRecord.toReference), name, attribute, refsToIds)
+
+              // if the attribute already exists it is an update, else an insert
+              if (existingAttributes.contains(name)) recordsForUpdateAttribute(name, attribute, attrRecords)
+              else (attrRecords, Seq.empty[EntityAttributeRecord], Seq.empty[Long])
+          }.foldLeft((Seq.empty[EntityAttributeRecord], Seq.empty[EntityAttributeRecord], Seq.empty[Long])) {
+            (t1, t2) => (t1, t2) match {
```

Comment:
What do the characters `t`, `i`, `u`, and `d` stand for? This is not clear to a reader IMO.

Reply:
t -> tuple, i-> inserts, u-> updates and d -> deletes. I shortened them to avoid the lengthy statements. But since is not clear to the reader I will change it.